### ISSUE TITLE
OOM error when stream is deployed/undeployed repeatedly

### ISSUE THREAD

**Issue by ilayaperumalg** [spring-cloud/spring-cloud-dataflow-server-cloudfoundry#107]
Please see the discussion here: https://github.com/spring-cloud/spring-cloud-dataflow-server-cloudfoundry/issues/93#issuecomment-221338070

**Comment by ilayaperumalg** [comment]
The OOM error might be related to  `async` handling of deploy in the CloudFoundry app deployer. When the deploy/undeploy operations are invoked continuously, then the following exception is thrown:

```
2016-05-25T16:12:29.22+0530 [APP/0]      OUT java.lang.IllegalStateException: Application a1-time failed during staging
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at org.cloudfoundry.util.ExceptionUtils.illegalState(ExceptionUtils.java:60) ~[cloudfoundry-util-2.0.0.M4.jar!/:2.0.0.M4]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at org.cloudfoundry.operations.applications.DefaultApplications.waitForStaging(DefaultApplications.java:1353) ~[cloudfoundry-operations-2.0.0.M4.jar!/:2.0.0.M4]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at org.cloudfoundry.operations.applications.DefaultApplications.lambda$startApplicationAndWait$102(DefaultApplications.java:1180) ~[cloudfoundry-operations-2.0.0.M4.jar!/:2.0.0.M4]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:371) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:125) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.MonoNext$NextSubscriber.onNext(MonoNext.java:85) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.UnicastProcessor.drain(UnicastProcessor.java:131) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.UnicastProcessor.onNext(UnicastProcessor.java:213) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.FluxSubscribeOn$SubscribeOnPipeline.onNext(FluxSubscribeOn.java:134) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.subscriber.SignalEmitter.emit(SignalEmitter.java:198) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.subscriber.SignalEmitter.onNext(SignalEmitter.java:444) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at org.cloudfoundry.spring.util.AbstractSpringOperations.lambda$null$1(AbstractSpringOperations.java:68) ~[cloudfoundry-client-spring-2.0.0.M4.jar!/:2.0.0.M4]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.FluxYieldingEmitter$YieldingSignalEmitter.request(FluxYieldingEmitter.java:75) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.util.DeferredSubscription.set(DeferredSubscription.java:71) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.FluxSubscribeOn$SubscribeOnPipeline.onSubscribe(FluxSubscribeOn.java:129) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.subscriber.SignalEmitter.start(SignalEmitter.java:170) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.FluxYieldingEmitter.subscribe(FluxYieldingEmitter.java:43) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:153) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.FluxFlatMap.subscribe(FluxFlatMap.java:170) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.FluxSubscribeOn$SourceSubscribeTask.run(FluxSubscribeOn.java:363) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.Computations$ProcessorWorker.onNext(Computations.java:919) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.Computations$ProcessorWorker.onNext(Computations.java:883) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at reactor.core.publisher.WorkQueueProcessor$QueueSubscriberLoop.run(WorkQueueProcessor.java:842) ~[reactor-core-2.5.0.M3.jar!/:na]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_71-]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_71-]
2016-05-25T16:12:29.22+0530 [APP/0]      OUT 	at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_71-]
```
and subsequently the OOM error occurs.

**Comment by ilayaperumalg** [comment]
One way to avoid this happening would be to make `deploy`/`undeploy` operations synchronous (waiting for the response from the CF deployer).

**Comment by sabbyanandan** [comment]
This is starting to occur all the more frequently now. As simple as the following stream and upon (2-3) repeated deploy/undeploy attempts crashes the server. 

> s3 | task-processor | task-laucnher

I'd think this can be reproduced with any of the OOTB apps.

Even after the stream is destroyed and all the apps are deleted, the memory consumption at the server level wouldn't go down. It tends to aggregate over repeated deployments and it eventually crashes.

**Comment by sabbyanandan** [comment]
The core problem was fixed in [`reactor-netty`](https://github.com/reactor/reactor-netty) and [`cf-java-client`](https://github.com/cloudfoundry/cf-java-client/commit/d3e6229b8dda4e67c5102db99ed55784ed6b206a) projects that the SCDF's CF-server relies upon for orchestrating streams/tasks in Cloud Foundry. Corresponding [cf-deployer](https://github.com/spring-cloud/spring-cloud-deployer-cloudfoundry/pull/116) and [cf-server](https://github.com/spring-cloud/spring-cloud-dataflow-server-cloudfoundry/pulls?q=is%3Apr+is%3Aclosed) PRs include the relevant updates. 

This will be released as part of the upcoming 1.1 RC release.
